### PR TITLE
Image to Image Generation Enabling

### PR DESCRIPTION
**Pull request by pi314ever**
# What does this PR do?

This model enables the base img2img model for Gaudi (other img2img pipelines were enabled, but not this base model). This also allows for chained image to image to image pipelines as shown in the [generative tasks for img2img docs](https://huggingface.co/docs/diffusers/en/using-diffusers/img2img#image-to-image-to-image). 

Validated models:
- [runwayml/stable-diffusion-v1-5](https://huggingface.co/runwayml/stable-diffusion-v1-5)
- [ogkalu/Comic-Diffusion](https://huggingface.co/ogkalu/Comic-Diffusion)
- [kohbanye/pixel-art-style](https://huggingface.co/kohbanye/pixel-art-style)

## Before submitting
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: tests/test_diffusers.py [hunk 1]
```diff
@@ -2421,6 +2422,144 @@ def test_stable_diffusion_pix2pix_euler(self):
         self.assertLess(np.abs(image_slice.flatten() - expected_slice).max(), 1e-3)
 
 
+class GaudiStableDiffusionImg2ImgPipelineTests(TestCase):
+    """
+    Tests the class StableDiffusionImg2ImgPipeline for Gaudi.
+    Adapted from: https://github.com/huggingface/diffusers/blob/v0.26.3/tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py
```
Comment (regisss) [inline tests/test_diffusers.py:2428]: Can you check if Diffusers 0.29.2 brings any changes?

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+    # Handles the case where the scheduler cannot implement reset_timestep_dependent_params()
+    # Example: UniPCMultiStepScheduler used for inference in ControlNet training as it has non-linear accesses to timestep dependent parameter: sigma.
+    if hasattr(scheduler, "reset_timestep_dependent_params") and callable(scheduler.reset_timestep_dependent_params):
+        scheduler.reset_timestep_dependent_params()
+    return timesteps, num_inference_steps
+
+
+class GaudiStableDiffusionImg2ImgPipeline(GaudiDiffusionPipeline, StableDiffusionImg2ImgPipeline):
+    """
+    Adapted from: https://github.com/huggingface/diffusers/blob/v0.26.3/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L161
```
Comment (regisss) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:102]: Can you check if Diffusers 0.29.2 brings any changes?

**Review by regisss — COMMENTED** [review]
I left a few minor comments. Can you rebase this branch on main to make sure the CI works?

File: tests/test_diffusers.py (hunk 1 above)
Comment (pi314ever) [reply tests/test_diffusers.py:2678]: No changes

**Comment by vidyasiv** [comment]
Hi @pi314ever ,

Thanks for your PR, please address:
- Run `make style` as code quality checker is failing
- Can you provide a command to test your code/ does the README need any update for this change?

**Review by vidyasiv — COMMENTED** [review]
Tests pass and also tested comic diffusion and pixel-art-style models with single and multi prompts from README. The runwayml/stable-diffusion-v1-5 no longer exists(https://github.com/huggingface/diffusers/issues/9322).
Comic diffusion seems to generate a few blank images- not sure why- do you also see this issue @pi314ever ?

**Comment by pi314ever** [comment]
@vidyasiv Could you share the prompts and source images of the blank images? I did not encounter those in my testing yet. That particular model seems to need `pixelartstyle` appended to the end of the prompt to work correctly according to the [example](https://huggingface.co/docs/diffusers/en/using-diffusers/img2img#image-to-image-to-image).

**Comment by vidyasiv** [comment]
> @vidyasiv Could you share the prompts and source images of the blank images? I did not encounter those in my testing yet. That particular model seems to need `pixelartstyle` appended to the end of the prompt to work correctly according to the [example](https://huggingface.co/docs/diffusers/en/using-diffusers/img2img#image-to-image-to-image).

I ran single and multiple prompt with https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion#single-prompt-1 changing model name to `ogkalu/Comic-Diffusion`. Will go through link you pasted.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+            # 7.2 Optionally get Guidance Scale Embedding
+            timestep_cond = None
+            if self.unet.config.time_cond_proj_dim is not None:
+                guidance_scale_tensor = torch.tensor(self.guidance_scale - 1).repeat(
+                    batch_size * num_images_per_prompt
+                )
+                timestep_cond = self.get_guidance_scale_embedding(
+                    guidance_scale_tensor, embedding_dim=self.unet.config.time_cond_proj_dim
+                ).to(device=device, dtype=latents.dtype)
+
```
Comment (vidyasiv) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:403]: Can you verify if we need _split_inputs_into_batches after guidance scale embedding similar to
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L473C75-L473C106
?

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+            # 8. Denoising loop
```
Comment (vidyasiv) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:404]: Might be useful to have Habana profiler (similar to
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L489
) if its quick to add

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+                        prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
+                        negative_prompt_embeds = callback_outputs.pop("negative_prompt_embeds", negative_prompt_embeds)
+
+                    # call the callback, if provided
+                    if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                        progress_bar.update()
+                        if callback is not None and i % callback_steps == 0:
+                            step_idx = i // getattr(self.scheduler, "order", 1)
+                            callback(step_idx, t, latents)
+
```
Comment (vidyasiv) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:457]: We can consider adding metrics measurement too (similar to
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L597
)

**Review by vidyasiv — CHANGES_REQUESTED** [review]
Thanks for code style update, its clean now. Adding some comments from my second pass of review

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 3 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:452]: Added

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 4 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py]: It looks relatively simple to add at first glance, I will try to integrate this into the PR.

Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py]: Done

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 5 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:559]: Done

File: tests/test_diffusers.py [hunk 6]
```diff
@@ -2676,6 +2677,144 @@ def test_stable_diffusion_pix2pix_euler(self):
...
+            "prompt": "A painting of a squirrel eating a burger",
+            "image": image,
+            "generator": generator,
+            "num_inference_steps": 2,
+            "guidance_scale": 6.0,
+            "output_type": "np",
+        }
+        return inputs
+
+    def test_stable_diffusion_img2img_default_case(self):
```
Comment (vidyasiv) [inline tests/test_diffusers.py:2759]: Reminder to update test values
```
FAILED tests/test_diffusers.py::GaudiStableDiffusionImg2ImgPipelineTests::test_stable_diffusion_img2img_default_case - AssertionError: assert 0.23589799033187864 < 0.001
FAILED tests/test_diffusers.py::GaudiStableDiffusionImg2ImgPipelineTests::test_stable_diffusion_img2img_multiple_init_images - AssertionError: assert 0.18246942101097108 < 0.001
FAILED tests/test_diffusers.py::GaudiStableDiffusionImg2ImgPipelineTests::test_stable_diffusion_img2img_negative_prompt - AssertionError: assert 0.24135875814361574 < 0.001
```

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 7]
```diff
@@ -0,0 +1,717 @@
...
+            outputs = {
+                "images": [],
+                "has_nsfw_concept": [],
+            }
+
+            t0 = time.time()
+            t1 = t0
+
+            hb_profiler = HabanaProfile(
```
Comment (vidyasiv) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:477]: Please verify profiling options:
```
python image_to_image_generation.py \
    --model_name_or_path "ogkalu/Comic-Diffusion"\
    --src_image_path "
[https://raw.githubusercontent.com/timothybrooks/instruct-pix2pix/main/imgs/example.jpg"](https://raw.githubusercontent.com/timothybrooks/instruct-pix2pix/main/imgs/example.jpg%22)
\
    --prompts "turn him into cyborg" \
    --num_images_per_prompt 20 \
    --profiling_warmup_steps 3 \
    --profiling_steps 2 \
    --batch_size 2 \
    --guidance_scale 7.5 \
    --image_guidance_scale 1 \
    --num_inference_steps 10 \
    --image_save_dir stable_diffusion_images \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --bf16
```
Error
```
Traceback (most recent call last):
  File "/root/PR/pi314ever_oh/examples/stable-diffusion/image_to_image_generation.py", line 338, in <module>
    main()
  File "/root/PR/pi314ever_oh/examples/stable-diffusion/image_to_image_generation.py", line 300, in main
    outputs = pipeline(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py", line 560, in __call__
    hb_profiler.step()
  File "/usr/local/lib/python3.10/dist-packages/torch/profiler/profiler.py", line 727, in step
    self._transit_action(prev_action, self.current_action)
  File "/usr/local/lib/python3.10/dist-packages/torch/profiler/profiler.py", line 744, in _transit_action
    action()
  File "/usr/local/lib/python3.10/dist-packages/habana_frameworks/torch/activity_profiler/activity_profiler.py", line 82, in start_trace
    super().start_trace()
  File "/usr/local/lib/python3.10/dist-packages/torch/profiler/profiler.py", line 160, in start_trace
    assert self.profiler is not None
AssertionError
```

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 8]
```diff
@@ -0,0 +1,717 @@
...
+            safety_checker,
+            feature_extractor,
+            image_encoder,
+            requires_safety_checker,
+        )
+        self.to(self._device)
+
+    # Copied from ./pipeline_stable_diffusion.py
+    @classmethod
+    def _split_inputs_into_batches(cls, batch_size, latents, prompt_embeds, negative_prompt_embeds):
```
Comment (vidyasiv) [inline optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:141]: Lower num_prompts_per_image is throwing error- could be we need some fix or input validation:
```
python image_to_image_generation.py \
    --model_name_or_path "ogkalu/Comic-Diffusion"\
    --src_image_path "
[https://raw.githubusercontent.com/timothybrooks/instruct-pix2pix/main/imgs/example.jpg"](https://raw.githubusercontent.com/timothybrooks/instruct-pix2pix/main/imgs/example.jpg%22)
\
    --prompts "turn him into cyborg" \
    --num_images_per_prompt 10 \
    --batch_size 4 \
    --guidance_scale 7.5 \
    --image_guidance_scale 1 \
    --num_inference_steps 10 \
    --image_save_dir stable_diffusion_images \
    --use_habana \
    --use_hpu_graphs \
    --gaudi_config Habana/stable-diffusion \
    --bf16
```
```
Traceback (most recent call last):
  File "/root/PR/pi314ever_oh/examples/stable-diffusion/image_to_image_generation.py", line 338, in <module>
    main()
  File "/root/PR/pi314ever_oh/examples/stable-diffusion/image_to_image_generation.py", line 300, in main
    outputs = pipeline(
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py", line 462, in __call__
    latents_batches, text_embeddings_batches, num_dummy_samples = self._split_inputs_into_batches(
  File "/usr/local/lib/python3.10/dist-packages/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py", line 180, in _split_inputs_into_batches
    prompt_embeds_batches = torch.stack(prompt_embeds_batches)
RuntimeError: stack expects each tensor to be equal size, but got [8, 77, 768] at entry 0 and [4, 77, 768] at entry 2
```

**Review by vidyasiv — CHANGES_REQUESTED** [review]
Almost there, added the 3 issues I found..

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 8 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:141]: Was caused by incorrect usage of `batch_size`, which was assumed to be `num_prompts` in the original. This was corrected.

File: tests/test_diffusers.py (hunk 6 above)
Comment (pi314ever) [reply tests/test_diffusers.py:2759]: Test outputs changed due to change in pipeline structure. Updated test expectations to match.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 7 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:477]: I will look into this tomorrow, as I am less familiar with how the profiler works.

**Comment by regisss** [comment]
@pi314ever There are merge conflicts to solve since I just merged #1175.

File: optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py (hunk 7 above)
Comment (pi314ever) [reply optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:477]: Fixed, just missed a `hb_profiler.start()`

**Review by vidyasiv — COMMENTED** [review]
LGTM

**Review by regisss — APPROVED** [review]

**Pull request merged**